app/features/settings/profile/editable-username.tsx

```diff
@@ -0,0 +1,153 @@
+import { Check, Edit } from 'lucide-react';
+import { useEffect, useRef, useState } from 'react';
+import { useForm } from 'react-hook-form';
+import { Button } from '~/components/ui/button';
+import { useUpdateUsername, useUser } from '~/features/user/user-hooks';
+import useAnimation from '~/hooks/use-animation';
+import { useToast } from '~/hooks/use-toast';
+import { cn } from '~/lib/utils';
+
+// Reserved routes that can't be used as usernames
+// QUESTION: can we do this more efficiently so that
+// we don't have to update for every new route?
+const RESERVED_ROUTES = [
+  'settings',
+  'profile',
+  'login',
+  'signup',
+  'api',
+  'admin',
+  'help',
+  'about',
+  'terms',
+  'privacy',
+  'receive',
+  'send',
+  'verify-email',
+];
+
+type FormValues = {
+  username: string;
+};
+
+function validateUsername(username: string) {
+  // Check length (3-20 characters)
+  if (username.length < 3 || username.length > 20) {
+    return 'Username must be between 3 and 20 characters';
+  }
+
+  // Check characters (only a-z, 0-9, underscore, hyphen allowed)
+  // This requirement comes from LUD16 https://github.com/lnurl/luds/blob/luds/16.md
+  if (!/^[a-z0-9_-]+$/.test(username)) {
```

Comment:
since our client side can write direclty to db we should probably have these checks done in the db as well right?

Reply:
Yes we can do it here as well